_posts/2017-03-05-artsy-technology-stack-2017.md

```diff
@@ -0,0 +1,156 @@
+
+We have a generic image-processing service in-house, which uses [Rails](http://rubyonrails.org), [Sidekiq](https://github.com/mperham/sidekiq/), [Redis](https://redis.io), and [RMagick](https://github.com/rmagick/rmagick) with [ImageMagick](http://www.imagemagick.org/script/index.php). It receives image processing requests from many Artsy applications and generates thumbnails, tiles and watermarks images on S3.
+
+## Collector-Facing
+
+Collectors inquire on artworks and engage in conversations with partners. For this purpose we have built a generic messaging system that manages communications between different parties. It receives messages via API or e-mail, finds or creates a conversation based on the recipients and forwards them to the proper addresses in that conversation. Its doesn't assume anything about the contents of the messages, which makes it a generic system for any type of conversation. The conversations surface to our partners via CMS.
+
+## Running Auctions
+
+Artsy's Auctions business started with charity auctions. Charity auctions are simpler to map digitally: they have less bids, are more free form in terms of bid increments, have less artworks for sale, they happen slower and the people running them are less risk-averse as they tend to be one-off annual events instead of regular occurrences. We therefore modeled these Auctions inside the core API. As we started implementing commercial auctions with a real-time component (we call these "Live Auctions") the differences in the scale of the domain made it worth moving the logic around auctions into a separate micro-service.
+
+The core API for a commercial auction is a Scala micro-service that uses [Akka](http://akka.io) for distributed computing. It stores information in an append-only storage engine, based on Akka Persistence, with a small library we developed called [atomic-store](https://github.com/artsy/atomic-store). Communication with external clients can either be done via a REST API, or via WebSockets powered by Akka Distributed Pub/Sub. People visiting a Live Auction on the web are interacting with a [universal](https://medium.com/@mjackson/universal-javascript-4761051b7ae9#.ev1yd3juy) [React](https://facebook.github.io/react)+[Redux](http://redux.js.org) JavaScript app, served from an [Express](http://expressjs.com) server. Bidders visiting a Live Auction on iOS are interacting with a Swift application built with [Interstellar](https://github.com/JensRavens/Interstellar), [Starscream](https://github.com/daltoniam/starscream) and [SwiftyJSON](https://github.com/SwiftyJSON/SwiftyJSON).
```

Comment:
I'd drop the Distributed Pub/Sub part. It is indeed connected with the Web Socket part, but it's more related to how state changes are broadcast across our Akka Cluster, so that they can then be forwarded on to active Web Socket subscribers. But that seems like it may be getting into the weeds a bit.